add a fact to cs_instance_facts. It contains all instance ip addresses split per network

fzwart opened the pull request:
##### SUMMARY
AFAIK there currently is no fact that queries the cloudstack API and returns all
IPs assigned to an instance (split per network). I've added this to cs_instance_facts

network_ip_addresses:
  <network_name>:
    - <ip1>
    - <ipN>

<!--- Describe the change, including rationale and design decisions -->
I felt this was the right place to add this functionality as the data is 
already there. It only needed to be added to the result. That's what I did.

##### ISSUE TYPE
- Feature Pull Request
 
##### COMPONENT NAME
cloud/cloudstack

##### ANSIBLE VERSION
```
ansible 2.6.2
  config file = None
  configured module search path = [u'/Users/frank/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Python/2.7/site-packages/ansible
  executable location = bin/ansible
  python version = 2.7.10 (default, Feb  7 2017, 00:08:15) [GCC 4.2.1 Compatible Apple LLVM 8.0.0 (clang-800.0.34)]
```
##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

webknjaz commented on lib/ansible/modules/cloud/cloudstack/cs_instance_facts.py:
```diff
@@ -265,9 +265,20 @@ def get_result(self, instance):
                     affinity_groups.append(affinitygroup['name'])
                 self.result['affinity_groups'] = affinity_groups
             if 'nic' in instance:
+                network_ip_addresses = {}
```
You could use `defaultdict(list)` here to avoid extra if clauses below.

webknjaz commented on lib/ansible/modules/cloud/cloudstack/cs_instance_facts.py:
```diff
@@ -265,9 +265,20 @@ def get_result(self, instance):
                         self.result['default_ip'] = nic['ipaddress']
+                    if 'ipaddress' in nic:
+                        network_ip_addresses[networkname].append(nic['ipaddress'])
+                    for secondaryip in nic['secondaryip']:
```
This would look more elegant:
```python
import itertools
...
for nic_section in itertools.chain((nic, ), nic['secondaryip']):
    try:
        ip_address = nic_section['ipaddress']
    except KeyError:
        continue
    network_ip_addresses[networkname].append(ip_address)
```

webknjaz replied:
BTW, if `secondaryip` might be missing, replace `nic['secondaryip']` with `nic.get('secondaryip', ())`

fzwart commented:
Nice feedback, thanks!
I've pushed both changes and squashed the commits, although I'm not sure if I should have done that as we now lost your feedback.

webknjaz commented on lib/ansible/modules/cloud/cloudstack/cs_instance_facts.py:
```diff
@@ -184,6 +185,8 @@
 
 from ansible.module_utils.basic import AnsibleModule
 from ansible.module_utils.cloudstack import AnsibleCloudStack, cs_argument_spec
+from collections import defaultdict
```
Please put these *stdlib* imports above project imports separating sections with an empty line.

webknjaz replied:
The convention is to group imports:
1) stdlib imports
2) common third-party imports
3) main framework imports (if applicable, not in this case)
4) own project imports (ansible, in our case)

webknjaz commented on lib/ansible/modules/cloud/cloudstack/cs_instance_facts.py:
```diff
@@ -265,9 +268,20 @@ def get_result(self, instance):
+                    networkname = nic['networkname']
                     if nic['isdefault'] and 'ipaddress' in nic:
                         self.result['default_ip'] = nic['ipaddress']
+                    for nic_section in itertools.chain((nic, ), nic['secondaryip']):
```
Could you please amend this commit adding `Co-authored-by: Sviatoslav Sydorenko <wk@sydorenko.org.ua>` trailed to long commit description?

webknjaz commented:
> as we now lost your feedback.

@fzwart that's not true. You can see it above (https://github.com/ansible/ansible/pull/44112#pullrequestreview-146058841), it's just collapsed. Click on "Show outdated" to see it.

fzwart committed:
Add the network_ip_addresses fact to cs_instance_facts.

It contains all ip addresses assigned to the instance.

network_ip_addresses:
  <network_name>:
    - <ip1>
    - <ipN>

Co-authored-by: Sviatoslav Sydorenko <wk@sydorenko.org.ua>

webknjaz reviewed (approved):
Good job!

webknjaz commented:
Now we need to wait for @dpassante @resmo to review this :)

dpassante commented on lib/ansible/modules/cloud/cloudstack/cs_instance_facts.py:
```diff
@@ -265,9 +268,20 @@ def get_result(self, instance):
+                            continue
+                        network_ip_addresses[networkname].append(ip_address)
+
+                self.result['network_ip_addresses'] = network_ip_addresses
```
Could you please add a description of ```network_ip_address``` in the [RETURN](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block) section of the module?

fzwart replied:
done

dpassante replied:
Thank's!

dpassante commented:
Thanks for your contribution ! (and sorry for the late reply)
Sounds good.

dpassante commented:
-label affects_2.7
+label affects_2.8

dpassante commented:
shipit

dpassante commented:
bot_status
